src/inmanta/agent/handler.py

```diff
@@ -822,6 +822,22 @@ def call() -> typing.Awaitable[Result]:
         except Exception:
             raise Exception("Unable to upload file to the server.")
 
+    def execute_discover_resource(self, ctx: HandlerContext, resource: resources.Resource):
+        ctx.info("execute_discover_resource")
+        self.pre(ctx, resource)
+        result = self.discover_resource(ctx)
+        if result:
+
+            def call():
+                client = protocol.Client("agent")
+                return client.discovered_resources_create(
```

Comment:
To add to this, to me this doesn't feel like it fits well with the current responsibilities of the handler. As it is now, the handler is an expert in its domain, but is nothing more than a slave to the agent in other respects, and it doesn't have side effects outside of its domain or the mutable parameters it receives. e.g. the agent asks the handler to deploy, it does, reports changes via the context, then the agent does all the bookkeeping (report facts, mark deploy done, ...). I think it might be more consistent to have this method report what it finds (possibly via the context?), and leave the processing/reporting to the agent. That way responsibilities are kept decoupled: the handler is the domain expert, the agent knows about the orchestration aspect.

Do you see what I mean? Wdyt?